timemachine/cpp/src/barostat.cu

```diff
@@ -172,41 +202,115 @@ void __global__ find_group_centroids(
     atomicAdd(centroids + mol_idx*3+2, FLOAT_TO_FIXED<double>(coords[atom_idx*3+2]));
 }
 
-void MonteCarloBarostat::reset_counters() {
-    num_attempted_ = 0;
-    num_accepted_ = 0;
+void __global__ k_setup_barostat_move(
+    const double * rand, // [2], use first value, second value is metropolis condition
+    double * __restrict__ d_box, // [3*3]
+    double * __restrict__ d_volume_delta, // [1]
+    double * __restrict__ d_volume_scale, // [1]
+    double * __restrict__ d_length_scale // [1]
+) {
+    const int idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (idx >= 1) {
+        return; // Only a single thread needs to perform this operation
+    }
+    const double volume = d_box[0*3+0] * d_box[1*3+1] * d_box[2*3+2];
+    if(d_volume_scale[0] == 0) {
+        d_volume_scale[0] = 0.01 * volume;
+    }
+    const double delta_volume = d_volume_scale[0]*2*(rand[0]-0.5);
+    const double new_volume = volume+delta_volume;
+    d_volume_delta[0] = delta_volume;
+    d_length_scale[0] = cbrt(new_volume/volume);
 }
 
-void MonteCarloBarostat::inplace_move(
-    double *d_x,
-    double *d_box,
-    const double lambda
-    ) {
+void __global__ k_decide_move(
+    const int N,
+    const int num_molecules,
+    const double kt,
+    const double pressure,
+    const double * __restrict__ rand, // [2] Use second value
+    double * __restrict__ d_volume_delta,
+    double * __restrict__ d_volume_scale,
+    const unsigned long long * __restrict__ d_init_u,
+    const unsigned long long * __restrict__ d_final_u,
+    double * __restrict__ d_box,
+    const double * __restrict__ d_box_output,
+    double * __restrict__ d_x,
+    const double * __restrict__ d_x_output,
+    int * __restrict__ num_accepted,
+    int * __restrict__ num_attempted
+) {
+    const int idx = blockIdx.x*blockDim.x + threadIdx.x;
+    if (idx >= N) {
+        return;
+    }
 
-    step_++;
-    if (step_ % interval_ != 0) {
+    const double volume = d_box[0*3+0] * d_box[1*3+1] * d_box[2*3+2];
+    const double new_volume = volume+d_volume_delta[0];
+    const double energy_delta = FIXED_TO_FLOAT<double>(d_final_u[0] - d_init_u[0]);
+    const double w = energy_delta + pressure*d_volume_delta[0] - num_molecules * kt * std::log(new_volume/volume);
+
+    const bool rejected = w > 0 && rand[1] > std::exp(-w/kt);
```

Comment:
Question: does the `w > 0` short-circuit optimization applied here help when this is done on the GPU? (Not sure if it's possible for this to introduce a slow down, since it introduces another branch...)

Reply:
I will have to check.

It seems like a nice additional to check as it adds a simple and explicit condition upon which the metropolis check is performed.

Reply:
This is easily short circuited. (also why you should never overload the comparison operators since short circuitting is "disabled" when you do).